e2e/mobile/utils/initUtil.ts

```diff
@@ -0,0 +1,271 @@
+        // Create fresh instance for next retry attempt
+        await deleteSpeculos(entry.deviceId);
+        const device = await launchSpeculos(app.name);
+
+        entryMap[app.name] = {
+          name: app.name,
+          speculosPort: device.port,
+          proxyPort: entry.proxyPort,
+          deviceId: device.id,
+        };
+      }
+    }
+  }
+
+  if (lastError) {
+    throw new Error(
+      `❌ [${app.name}] Failed to setup account after ${maxRetries} attempts: ${lastError}`,
+    );
+  }
+
+  await deleteSpeculos(entry.deviceId);
+}
+
+// Execute commands for each app with retry mechanism
+async function executeCliCommandsOnApp(
+  uniqueOnApp: Array<{ app: SpeculosAppType; cmd: CliCommand }>,
+  entryMap: Record<string, Entry>,
+  userdataPath: string,
+): Promise<void> {
+  for (const { app, cmd } of uniqueOnApp) {
+    await executeAppCommandWithRetry(app, cmd, entryMap, userdataPath);
+  }
+}
+
+// Retry logic for main Speculos app setup with instance recreation
+async function setupMainSpeculosApp(
+  speculosApp: SpeculosAppType,
+  entryMap: Record<string, Entry>,
+): Promise<void> {
+  const main = entryMap[speculosApp.name];
+  if (!main) {
+    throw new Error(`No entry found for main speculos app: ${speculosApp.name}`);
+  }
+
+  const maxRetries = 3;
+  let attempt = 0;
+  let lastError: unknown;
+
+  while (attempt < maxRetries) {
+    attempt++;
+
+    try {
+      log.info(`\n🔄 [${speculosApp.name}] Main setup attempt ${attempt}/${maxRetries}`);
+
+      await waitForSpeculosIfRemote(main);
+      await registerSpeculos(main.speculosPort, main.proxyPort);
+      await registerKnownSpeculos(main.proxyPort);
+      log.info(
+        `✅ [${speculosApp.name}] Main Speculos registered successfully on port ${main.speculosPort}`,
+      );
+
+      lastError = undefined;
+      break;
+    } catch (err) {
+      lastError = err;
+
+      if (attempt < maxRetries) {
+        log.info(`[${speculosApp.name}] Creating new main Speculos instance for retry`);
+        await removeSpeculosAndDeregisterKnownSpeculos(main.deviceId);
+        const device = await launchSpeculos(main.name);
+
+        entryMap[speculosApp.name] = {
+          name: main.name,
+          speculosPort: device.port,
+          proxyPort: main.proxyPort,
+          deviceId: device.id,
+        };
+      }
+    }
+  }
+
+  if (lastError) {
+    throw new Error(
+      `❌ [${speculosApp.name}] Failed to setup main Speculos app after ${maxRetries} attempts: ${lastError}`,
+    );
+  }
+}
+
+// Execute global commands after all app-specific setup is complete
+async function executeCliCommands(
+  cliCommands: CliCommand[],
+  userdataPath: string,
+  speculosApp?: SpeculosAppType,
+  entryMap?: Record<string, Entry>,
+): Promise<void> {
+  for (const cmd of cliCommands) {
+    if (speculosApp && isRemoteIos() && entryMap) {
+      const main = entryMap[speculosApp.name];
+      await waitForSpeculosReady(getSpeculosAddress(main.deviceId));
```

Comment:
Why do we need to wait for readiness as we already did on `setupMainSpeculosApp` ?